compiler/rustc_trait_selection/src/solve/eval_ctxt/select.rs

```diff
@@ -25,359 +20,194 @@ impl<'tcx> InferCtxt<'tcx> {
-
-                (Certainty::Maybe(_), _) => Ok(None),
             }
-        })
-    }
-}
+        }
 
-impl<'tcx> EvalCtxt<'_, 'tcx> {
-    fn compute_canonical_trait_candidates(
-        &mut self,
-        canonical_input: CanonicalInput<'tcx>,
-    ) -> Vec<Candidate<'tcx>> {
-        // This doesn't record the canonical goal on the stack during the
-        // candidate assembly step, but that's fine. Selection is conceptually
-        // outside of the solver, and if there were any cycles, we'd encounter
-        // the cycle anyways one step later.
-        EvalCtxt::enter_canonical(
-            self.tcx(),
-            self.search_graph,
-            canonical_input,
-            // FIXME: This is wrong, idk if we even want to track stuff here.
-            &mut ProofTreeBuilder::new_noop(),
-            |ecx, goal| {
-                let trait_goal = Goal {
-                    param_env: goal.param_env,
-                    predicate: goal
-                        .predicate
-                        .to_opt_poly_trait_pred()
-                        .expect("we canonicalized a trait goal")
-                        .no_bound_vars()
-                        .expect("we instantiated all bound vars"),
-                };
-                ecx.assemble_and_evaluate_candidates(trait_goal)
-            },
-        )
+        ControlFlow::Break(Ok(to_selection(self.span, candidates.into_iter().next().unwrap())))
     }
 }
 
+/// This is a lot more limited than the old solver's equivalent method. This may lead to more `Ok(None)`
+/// results when selecting traits in polymorphic contexts, but we should never rely on the lack of ambiguity,
+/// and should always just gracefully fail here. We shouldn't rely on this incompleteness.
 fn candidate_should_be_dropped_in_favor_of<'tcx>(
-    tcx: TyCtxt<'tcx>,
-    victim: &Candidate<'tcx>,
-    other: &Candidate<'tcx>,
+    victim: &inspect::InspectCandidate<'_, 'tcx>,
+    other: &inspect::InspectCandidate<'_, 'tcx>,
 ) -> bool {
-    match (victim.source, other.source) {
-        (CandidateSource::ParamEnv(victim_idx), CandidateSource::ParamEnv(other_idx)) => {
-            victim_idx >= other_idx
-        }
-        (_, CandidateSource::ParamEnv(_)) => true,
+    // Don't winnow until `Certainty::Yes` -- we don't need to winnow until
+    // codegen, technically.
+    if matches!(other.result().unwrap(), Certainty::Maybe(..)) {
+        return false;
+    }
 
-        // FIXME: we could prefer earlier vtable bases perhaps...
+    match (victim.kind(), other.kind()) {
+        // Prefer dyn candidates over non-dyn candidates. This is necessary to
+        // handle the unsoundness between `impl<T: ?Sized> Any for T` and `dyn Any: Any`.
         (
-            CandidateSource::BuiltinImpl(BuiltinImplSource::Object { .. }),
-            CandidateSource::BuiltinImpl(BuiltinImplSource::Object { .. }),
+            inspect::ProbeKind::TraitCandidate {
```

Comment:
please first unwrap the `ProbeKind::TraitCandidate` of both `victim` and `other`

Reply:
what do you mean unwrap?

Reply:
```rs
let inspect::ProbeKind::TraitCandidate { source: victim_source, result: _ } = victim.kind() else { return false; };
let inspect::ProbeKind::TraitCandidate { source: other_source, result: _ } = other.kind() else { return false; };
```

Reply:
```rs
let inspect::ProbeKind::TraitCandidate { source: victim_source, result: _ } = victim.kind() else { return false; };
let inspect::ProbeKind::TraitCandidate { source: other_source, result: _ } = other.kind() else { return false; };

match (victim_source, other_source) {
    // Prefer dyn candidates over non-dyn candidates. This is necessary to
    // handle the unsoundness between `impl<T: ?Sized> Any for T` and `dyn Any: Any`.
    (
        CandidateSource::BuiltinImpl(BuiltinImplSource::Object { .. }),
        CandidateSource::BuiltinImpl(BuiltinImplSource::Object { .. }),
    ) => false,

    (
        _,
        CandidateSource::BuiltinImpl(BuiltinImplSource::Object { .. }),
    ) => false,

    // Prefer specializing candidates over specialized candidates.
    (
        CandidateSource::Impl(victim_def_id),
        CandidateSource::Impl(other_def_id),
    ) => victim.goal().infcx().tcx.specializes((other_def_id, victim_def_id)),

    _ => false,
}
```